ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -0,0 +1,487 @@
+  private final HelixClusterManager.ClusterChangeHandlerCallback clusterChangeHandlerCallback;
+  private final CountDownLatch routingTableInitLatch = new CountDownLatch(1);
+  // A map whose key is ambry datanode and value is a map of partitionName to corresponding replica on this datanode
+  // Note: the partitionName (inner map key) comes from partitionId.toString() not partitionId.toPathString()
+  private final ConcurrentHashMap<AmbryDataNode, ConcurrentHashMap<String, AmbryReplica>> ambryDataNodeToAmbryReplicas =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<AmbryDataNode, Set<AmbryDisk>> ambryDataNodeToAmbryDisks = new ConcurrentHashMap<>();
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  private volatile ConcurrentHashMap<String, String> partitionNameToResource = new ConcurrentHashMap<>();
+  private AtomicReference<RoutingTableSnapshot> routingTableSnapshotRef = new AtomicReference<>();
+  private ConcurrentHashMap<String, AmbryDataNode> instanceNameToAmbryDataNode = new ConcurrentHashMap<>();
+  private static final Logger logger = LoggerFactory.getLogger(DynamicClusterChangeHandler.class);
+
+  DynamicClusterChangeHandler(ClusterMapConfig clusterMapConfig, String dcName, String selfInstanceName,
+      Map<String, Map<String, String>> partitionOverrideInfoMap,
+      HelixClusterManager.HelixClusterManagerCallback helixClusterManagerCallback,
+      HelixClusterManager.ClusterChangeHandlerCallback clusterChangeHandlerCallback,
+      HelixClusterManagerMetrics helixClusterManagerMetrics,
+      ConcurrentHashMap<String, Exception> initializationFailureMap, AtomicLong sealedStateChangeCounter) {
+    this.clusterMapConfig = clusterMapConfig;
+    this.dcName = dcName;
+    this.selfInstanceName = selfInstanceName;
+    this.partitionOverrideInfoMap = partitionOverrideInfoMap;
+    this.helixClusterManagerCallback = helixClusterManagerCallback;
+    this.clusterChangeHandlerCallback = clusterChangeHandlerCallback;
+    this.helixClusterManagerMetrics = helixClusterManagerMetrics;
+    this.initializationFailureMap = initializationFailureMap;
+    this.sealedStateChangeCounter = sealedStateChangeCounter;
+  }
+
+  /**
+   * Handle any {@link InstanceConfig} related change in current datacenter. Several events will trigger instance config
+   * change: (1) replica's seal or stop state has changed; (2) new node or new partition is added; (3) new replica is
+   * added to existing node; (4) old replica is removed from existing node; (5) data node is deleted from cluster.
+   * For now, {@link DynamicClusterChangeHandler} supports (1)~(4). We may consider supporting (5) in the future.
+   * @param configs all the {@link InstanceConfig}(s) in current data center. (Note that PreFetch is enabled by default
+   *                in Helix, which means all instance configs under "participants" ZNode will be sent to this method)
+   * @param changeContext the {@link NotificationContext} associated.
+   */
```

Comment:
can we add some comments with regards to the znode path in the zookeeper? I think other engineers would benefit a lot from this information.

Reply:
sure, will do